Overview: Not compiling with waf

Overview plugin is not compiling with waf

    //.....
    Summary: 
    Install Geany Plugins 1.25 in            : /usr/local 
    Using GTK version                        : 2.24.25 
    Using Geany version                      : 1.25 
    Compiling Git revision                   : 8f70319 
    Plugins to compile                       : overview 
    Plugins to skip due to missing dependencies :  
    'configure' finished successfully (0.075s)
    Waf: Entering directory `/home/frlan/quellen/geany-plugins/_build_'
    Traceback (most recent call last):
      File "/home/frlan/quellen/geany-plugins/.waf-1.6.11-30618c54883417962c38f5d395f83584/waflib/Scripting.py", line 97, in waf_entry_point
        run_commands()
      File "/home/frlan/quellen/geany-plugins/.waf-1.6.11-30618c54883417962c38f5d395f83584/waflib/Scripting.py", line 151, in run_commands
        run_command(cmd_name)
      File "/home/frlan/quellen/geany-plugins/.waf-1.6.11-30618c54883417962c38f5d395f83584/waflib/Scripting.py", line 143, in run_command
        ctx.execute()
      File "/home/frlan/quellen/geany-plugins/.waf-1.6.11-30618c54883417962c38f5d395f83584/waflib/Scripting.py", line 345, in execute
        return execute_method(self)
      File "/home/frlan/quellen/geany-plugins/.waf-1.6.11-30618c54883417962c38f5d395f83584/waflib/Build.py", line 98, in execute
        self.execute_build()
      File "/home/frlan/quellen/geany-plugins/.waf-1.6.11-30618c54883417962c38f5d395f83584/waflib/Build.py", line 101, in execute_build
        self.recurse([self.run_dir])
      File "/home/frlan/quellen/geany-plugins/.waf-1.6.11-30618c54883417962c38f5d395f83584/waflib/Context.py", line 128, in recurse
        user_function(self)
      File "/home/frlan/quellen/geany-plugins/wscript", line 254, in build
        bld.recurse(enabled_plugins)
      File "/home/frlan/quellen/geany-plugins/.waf-1.6.11-30618c54883417962c38f5d395f83584/waflib/Context.py", line 112, in recurse
        exec(compile(function_code,node.abspath(),'exec'),self.exec_dict)
      File "/home/frlan/quellen/geany-plugins/overview/wscript_build", line 36, in <module>
        defines = [ subst_vars('OVERVIEW_PREFS_UI_FILE="${PKGDATADIR}/overview/prefs.ui"', bld.env) ]
    NameError: name 'subst_vars' is not defined

Okay the error outlined here is easy to fix (just import the right thing).  But then you'll see you also need to remove the "packages" references -- which aren't valid for C plugins.  Both are fixed by 454b14e44c763d123457594d7fd3ab84598837fe

And then, you'll stumble on something more subtle: this plugin uses some C99-isms that aren't part of GNU89 (GCC's default), meaning it requires explicitly requesting C99 or newer, which Autotools does since b5ad2878205147302f4c60a16aae57786723a96b but Waf doesn't.  This leads to build failure:
```
[ 47/281] c: overview/overview/overviewui.c -> _build_/overview/overview/overviewui.c.1.o
../overview/overview/overviewui.c: In function ‘overview_ui_get_menu_item_pos’:
../overview/overview/overviewui.c:261:3: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
   for (GList *iter = children; iter != NULL; iter = g_list_next (iter), pos++)
   ^
../overview/overview/overviewui.c:261:3: note: use option -std=c99, -std=gnu99, -std=c11 or -std=gnu11 to compile your code
Waf: Leaving directory `/home/ban/progs/geany-plugins/_build_'
Build failed
 -> task in 'overview' failed (exit status 1): 
	{task 140306111322128: c overviewui.c -> overviewui.c.1.o}
['/home/ban/bin/gcc', '-Wall', '-W', '-Werror-implicit-function-declaration', '-Wno-unused-parameter', '-DHAVE_CONFIG_H', '-O2', '-g', '-DDEBUG', '-fPIC', '-pthread', '-pthread', '-pthread', '-pthread', '-I/home/ban/progs/geany-plugins/_build_/overview/overview', '-I/home/ban/progs/geany-plugins/overview/overview', '-I/home/ban/progs/geany-plugins/_build_', '-I/usr/include/gtk-2.0', '-I/usr/lib/x86_64-linux-gnu/gtk-2.0/include', '-I/usr/include/gio-unix-2.0', '-I/usr/include/cairo', '-I/usr/include/pango-1.0', '-I/usr/include/atk-1.0', '-I/usr/include/pixman-1', '-I/usr/include/libpng12', '-I/usr/include/gdk-pixbuf-2.0', '-I/usr/include/harfbuzz', '-I/usr/include/glib-2.0', '-I/usr/lib/x86_64-linux-gnu/glib-2.0/include', '-I/usr/include/freetype2', '-I/usr/local/include/geany', '-I/usr/local/include/geany/tagmanager', '-I/usr/local/include/geany/scintilla', '-I/usr/include/gtk-3.0', '-I/usr/include/at-spi2-atk/2.0', '-I/usr/include/at-spi-2.0', '-I/usr/include/dbus-1.0', '-I/usr/lib/x86_64-linux-gnu/dbus-1.0/include', '-DOVERVIEW_PREFS_UI_FILE="/usr/local/share/geany-plugins/overview/prefs.ui"', '-DG_LOG_DOMAIN="Overview"', '-DGTK', '../overview/overview/overviewui.c', '-c', '-o', 'overview/overview/overviewui.c.1.o']
```

--- a/build/wafutils.py
+++ b/build/wafutils.py
@@ -179,6 +179,67 @@ def check_cfg_cached(conf, **kw):
     return result
 
 
+def check_c99(conf):
+    key = 'c99'
+    if key in cache:
+        v = cache[key]
+        if v:
+            raise v
+        return True
+
+    # FIXME: improve some checks?
+    # TODO: look at Autoconf's C99 checks?
+    fragment = '''
+    // single-line comments
+
+    #include <stdbool.h>
+
+    struct s { int a, b; };
+
+    // inlines
+    static inline void fun_inline(struct s param) {}
+
+    int main(void) {
+        _Bool b = false;
+
+        // variable declaration in for body
+        for (int i = 0; i < 2; i++);
+
+        // compound literals
+        fun_inline((struct s) { 1, 2 });
+
+        // mixed declarations and code
+        int mixed = 0;
+
+        // named initializers
+        struct s name_inited = {
+            .a = 42,
+            .b = 64
+        };
+
+        return (b || mixed || ! name_inited.a);
+    }
+    '''
+
+    exc = None
+    # list of flags is stolen from Autoconf 2.69
+    flags = ['', '-std=gnu99', '-std=c99', '-c99', '-AC99',
+             '-D_STDC_C99=', '-qlanglvl=extc99']
+    for flag in flags:
+        try:
+            desc = ['with flag %s' % flag, 'with no flags'][not flag]
+            conf.check_cc(fragment=fragment, uselib_store='C99', cflags=flag,
+                          msg="Checking for C99 support (%s)" % desc)
+            exc = None
+            break
+        except ConfigurationError as e:
+            exc = e
+    cache[key] = exc
+    if exc:
+        raise exc
+    return True
+
+
 def get_enabled_plugins(conf):
     plugins = get_plugins()
     enabled_plugins = []

--- a/overview/wscript_build
+++ b/overview/wscript_build
@@ -20,6 +20,7 @@
 #
 
 from build.wafutils import build_plugin
+from waflib.Utils import subst_vars
 
 name = 'Overview'
 sources = [
@@ -31,13 +32,11 @@ sources = [
   'overview/overviewui.c',
 ]
 includes = ['overview']
-packages = [ 'gtk+-2.0', 'glib-2.0', 'geany' ]
 libraries = [ 'GTK', 'GLIB', 'GEANY' ]
 defines = [ subst_vars('OVERVIEW_PREFS_UI_FILE="${PKGDATADIR}/overview/prefs.ui"', bld.env) ]
 
 build_plugin(bld, name,
              sources=sources,
              includes=includes,
-             packages=packages,
              libraries=libraries,
              defines=defines)

--- a/overview/wscript_build
+++ b/overview/wscript_build
@@ -32,7 +32,7 @@ sources = [
   'overview/overviewui.c',
 ]
 includes = ['overview']
-libraries = [ 'GTK', 'GLIB', 'GEANY' ]
+libraries = [ 'C99', 'GTK', 'GLIB', 'GEANY' ]
 defines = [ subst_vars('OVERVIEW_PREFS_UI_FILE="${PKGDATADIR}/overview/prefs.ui"', bld.env) ]
 
 build_plugin(bld, name,

--- a/overview/wscript_configure
+++ b/overview/wscript_configure
@@ -1,0 +1,3 @@
+from build.wafutils import check_c99
+
+check_c99(conf)

